# **Linear Regression Assignment**: California Housing Price Prediction

### **Part 1**: Data Loading, Preprocessing, and Initial Linear Regression

**First Improt the necessary Libraries**

In [60]:
import numpy as np
import pandas as pd
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

**Task 1.1**: Load and Prepare the Data

In [61]:
# Load California housing dataset
df = fetch_california_housing(as_frame=True)
X = df.data
y = df.target

In [62]:
# Dataset characteristics
data = fetch_california_housing()
print(data.DESCR)

.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

:Number of Instances: 20640

:Number of Attributes: 8 numeric, predictive attributes and the target

:Attribute Information:
    - MedInc        median income in block group
    - HouseAge      median house age in block group
    - AveRooms      average number of rooms per household
    - AveBedrms     average number of bedrooms per household
    - Population    block group population
    - AveOccup      average number of household members
    - Latitude      block group latitude
    - Longitude     block group longitude

:Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived from the 1990 U.S. census, using one row per ce

In [63]:
# First 5 rows of features
X.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

In [64]:
# First 5 rows
y.head()

0    4.526
1    3.585
2    3.521
3    3.413
4    3.422
Name: MedHouseVal, dtype: float64

**Examine the features and the target variable**

In [65]:
print(f"Feature dataset's shape: {X.shape}\n")
X.info()

Feature dataset's shape: (20640, 8)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
dtypes: float64(8)
memory usage: 1.3 MB


In [66]:
print(f"\nTarget Variable y Shape: {y.shape} Dtypes: {y.dtypes}")


Target Variable y Shape: (20640,) Dtypes: float64


**Want to see the Statistics of Features variables**

In [67]:
X.describe()

MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25%        2.429741     33.930000   -121.800000  
50%        2.818116     34.260000   -118.490000  
75%        3.282261     37.710000   -118.010000  
max     1243.333333     41.950000   -114.310000

**Split the data into training and testing sets (e.g., 80% train, 20% test)**

In [68]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)
print(f"\nShapes -> X_train: {X_train.shape}, X_test: {X_test.shape}, "
      f"y_train: {y_train.shape}, y_test: {y_test.shape}")


Shapes -> X_train: (16512, 8), X_test: (4128, 8), y_train: (16512,), y_test: (4128,)


**Standardize the feature data (fit on train, transform on both)**

In [69]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

**Task 1.2**: Train the Simple Linear Regression Model

In [70]:
model = LinearRegression()
model.fit(X_train_scaled, y_train)

LinearRegression()

In [71]:
# Predict on the test set
y_pred = model.predict(X_test_scaled)

**Task 1.3**: Evaluate the Simple Linear Regression Model

In [72]:
# Calculate evaluation metrics
mae  = mean_absolute_error(y_test, y_pred)
mse  = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2   = r2_score(y_test, y_pred)

# Print metrics
print("\n📈 Model Evaluation Metrics (Test Set):")
print(f"MAE  : {mae:.4f} (~${mae * 100000:,.0f})")
print(f"MSE  : {mse:.4f}")
print(f"RMSE : {rmse:.4f} (~${rmse * 100000:,.0f})")
print(f"R²   : {r2:.4f}")


📈 Model Evaluation Metrics (Test Set):
MAE  : 0.5332 (~$53,320)
MSE  : 0.5559
RMSE : 0.7456 (~$74,558)
R²   : 0.5758


**MAE (Mean Absolute Error)** : MAE represents the average absolute difference between predicted and actual values. It treats all errors equally and is easy to interpret, making it suitable when all prediction mistakes have similar importance. However, it's not ideal when large errors matter, as it does not penalize them more heavily.

**MSE (Mean Squared Error)** : MSE is the average of the squared differences between predicted and actual values. It penalizes larger errors more heavily than smaller ones, making it useful when big mistakes are especially undesirable. However, it’s less interpretable because the error is in squared units, not the same as the original target variable.

**RMSE (Root Mean Squared Error)** : RMSE is the square root of MSE, bringing the error back to the same unit as the target variable. Like MSE, it still penalizes large errors, but it's easier to understand in context. It's widely used when we care about both interpretability and sensitivity to big errors.

**R² (Coefficient of Determination)** : R² measures how well the model explains the variance in the target variable. It ranges from 0 to 1, where 1 means perfect prediction and 0 means the model does no better than the average.

### **Findings from simple linear regression mode** :

**MAE (Mean Absolute Error)** : In our model, MAE = 0.5332 (~$53,320). So this means that on average the model's predicted hosuing prices deviate from actual prices by 53K dollar.

**MSE (Mean Squared Error)** : In our model, MSE = 0.5559 (~$55,590²)
This indicates the average squared error in the predictions. This is not effective to interpret but sometimes useful because it will tell us large errors are present in our predictions

**RMSE (Root Mean Squared Error)** : In our model, RMSE = 0.7456 (~$74,560).
This means that, on average, the magnitude of error in our predictions is around 74K Dollar from the actual housing price, which reflects both typical and large errors more clearly than MAE.

**R² (Coefficient of Determination)** : R²: 0.5758 (~57.58%) Which indicates that the model explains 57.58% of the variability in house prices. In other words, almost 43% of the variation is still unexplained by the model

## **Part 2**: Regularized Linear Models (Lasso and Ridge)

**Brief Explanation of Ridge vs Lasso** :

**Ridge Regression** applies L2 regularization, which shrinks all coefficients, but none become exactly zero. It reduces overfitting by smoothing.

**Lasso Regression** applies L1 regularization, which can shrink some coefficients to exactly zero, effectively performing feature selection.

**Task 2.1**: Implement Ridge Regression

In [73]:
# Importing necessary libraries
from sklearn.linear_model import Ridge, Lasso
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

In [74]:
# Instantiate Ridge Regression model
ridge_model = Ridge(alpha=1.0)
ridge_model.fit(X_train_scaled, y_train)

# Predict on the test set
ridge_preds = ridge_model.predict(X_test_scaled)

Evaluate the Ridge Regression Model

In [75]:
# Evaluation metrics
ridge_mae  = mean_absolute_error(y_test, ridge_preds)
ridge_mse  = mean_squared_error(y_test, ridge_preds)
ridge_rmse = np.sqrt(ridge_mse)
ridge_r2   = r2_score(y_test, ridge_preds)

# Print results
print("📘 Ridge Regression Evaluation:")
print(f"MAE  : {ridge_mae:.4f} (~${ridge_mae * 100000:,.0f})")
print(f"MSE  : {ridge_mse:.4f}")
print(f"RMSE : {ridge_rmse:.4f} (~${ridge_rmse * 100000:,.0f})")
print(f"R²   : {ridge_r2:.4f}")

📘 Ridge Regression Evaluation:
MAE  : 0.5332 (~$53,319)
MSE  : 0.5559
RMSE : 0.7456 (~$74,556)
R²   : 0.5758


**Comments** : At alpha = 1.0, Ridge does not improve performance over simple linear regression on this dataset. The R-Square value of Simple linear regression and Ridge regression is similar in this case. Therefore, testing other alpha values (like 0.1, 10, 50, 100) is important to understand how regularization strength affects model performance

In [80]:
from sklearn.linear_model import Ridge

alphas = [0.01, 0.1, 1.0, 10, 50, 100]
print(f"{'Alpha':<10} {'R²':<10} {'RMSE':<10}")

for a in alphas:
    model = Ridge(alpha=a)
    model.fit(X_train_scaled, y_train)
    preds = model.predict(X_test_scaled)

    r2 = r2_score(y_test, preds)
    rmse = np.sqrt(mean_squared_error(y_test, preds))

    print(f"{a:<10} {r2:.4f}     {rmse:.4f}")

Alpha      R²         RMSE      
0.01       0.5758     0.7456
0.1        0.5758     0.7456
1.0        0.5758     0.7456
10         0.5761     0.7453
50         0.5770     0.7445
100        0.5778     0.7438


**Findings** : After testing multiple alpha values in Ridge regression, we observed that the model performance remained nearly identical to that of standard linear regression. This suggests that the California Housing dataset does not suffer from overfitting or high variance. Therefore, regularization does not provide significant benefit in this case

**Task 2.2**: Implement Lasso Regression

In [77]:
# Instantiate Lasso Regression model
lasso_model = Lasso(alpha=0.01)
lasso_model.fit(X_train_scaled, y_train)

# Make predictions
lasso_preds = lasso_model.predict(X_test_scaled)

Evaluate the Lasso Regression Model

In [81]:
# Evaluation metrics
lasso_mae  = mean_absolute_error(y_test, lasso_preds)
lasso_mse  = mean_squared_error(y_test, lasso_preds)
lasso_rmse = np.sqrt(lasso_mse)
lasso_r2   = r2_score(y_test, lasso_preds)

# Print results
print("\n📘 Lasso Regression Evaluation:")
print(f"MAE  : {lasso_mae:.4f} (~${lasso_mae * 100000:,.0f})")
print(f"MSE  : {lasso_mse:.4f}")
print(f"RMSE : {lasso_rmse:.4f} (~${lasso_rmse * 100000:,.0f})")
print(f"R²   : {lasso_r2:.4f}")


📘 Lasso Regression Evaluation:
MAE  : 0.5353 (~$53,533)
MSE  : 0.5483
RMSE : 0.7404 (~$74,044)
R²   : 0.5816


**Findings** : All three models give very similar performance on the test set, with Lasso slightly outperforming the others on R² and RMSE — but the difference is marginal.

Lasso Coefficient Analysis

In [82]:
print("\n🔍 Lasso Coefficients:")
for name, coef in zip(X.columns, lasso_model.coef_):
    print(f"{name:20s} : {coef:.4f}")


🔍 Lasso Coefficients:
MedInc               : 0.8010
HouseAge             : 0.1271
AveRooms             : -0.1628
AveBedrms            : 0.2062
Population           : -0.0000
AveOccup             : -0.0306
Latitude             : -0.7901
Longitude            : -0.7557


**Findings** : Lasso shrank the Population feature's coefficient to exactly 0, effectively removing it from the model. This demonstrates Lasso’s ability to perform feature selection, even when only 1 feature is removed with this small alpha (α = 0.01).

We can try different alpha values for Lasso as well to observe if anyhow we can see the improvement in the R-square value and if more features get zeroed.

In [84]:
from sklearn.linear_model import Lasso

alphas = [0.001, 0.01, 0.1, 0.5]
print(f"{'Alpha':<10} {'R²':<10} {'Zero Coeffs'}")

for a in alphas:
    model = Lasso(alpha=a, max_iter=10000)
    model.fit(X_train_scaled, y_train)
    preds = model.predict(X_test_scaled)

    r2 = r2_score(y_test, preds)
    zero_coeffs = np.sum(model.coef_ == 0)

    print(f"{a:<10} {r2:.4f}     {zero_coeffs}")

Alpha      R²         Zero Coeffs
0.001      0.5769     0
0.01       0.5816     1
0.1        0.4814     5
0.5        0.2827     7


### **Findings** : α = 0.01 gave the best performance (highest R² = 0.5816). Also eliminated 1 feature (Population) that was least helpful. but as As alpha increased, R² dropped significantly and more coefficients became zero which is like over regularization. So In conclusion Lasso is most useful when we expect many irrelevant features, but here the dataset is already clean and low-dimensional (only 8 features).

## **Part 3**: Analysis and Observation

**Task 3**: Final Comparison and Conclusion

| Model                           | R² Score | RMSE   | MAE    |
| ------------------------------- | -------- | ------ | ------ |
| **Simple Linear Regression**    | 0.5758   | 0.7456 | 0.5332 |
| **Lasso Regression (α = 0.01)** | 0.5816   | 0.7404 | 0.5353 |
| **Ridge Regression (α = 1.0)**  | 0.5758   | 0.7456 | 0.5332 |


### **Conclusion**

**Best Performing Model**:

Based on the evaluation metrics, the Lasso Regression model (with α = 0.01) performs the best for predicting median house values in California. It achieved:

The highest R² score (0.5816), meaning it explained the most variance in the target variable.

The lowest RMSE (0.7404), indicating the smallest average prediction error.

A comparable MAE (0.5353) to other models, which further confirms its reliability.

Although all three models performed similarly, Lasso had a slight edge in accuracy, making it the preferred model in this case.

### **Role of Regularization in Preventing Overfitting**

Regularization helps prevent overfitting by adding a penalty to large coefficients in the model:

**Ridge Regression** (L2 penalty) shrinks all coefficients toward zero, but doesn’t make any of them exactly zero. It helps in situations with multicollinearity or high variance, but in this dataset, it had little effect because the model wasn’t overfitting to begin with.

**Lasso Regression** (L1 penalty) can shrink some coefficients exactly to zero, effectively performing feature selection. In your analysis, Lasso eliminated the Population feature, simplifying the model without sacrificing performance.

In summary, regularization controls model complexity, especially when working with noisy or high-dimensional datasets. In our case, Lasso improved performance slightly and reduced complexity, even though overfitting wasn’t a major issue in the dataset.